Send email notification with build status, repository link and cause of build

navidshaikh opened the pull request:
This PR implements sending email notifications to users with
 - Status of build
 - Repository link (which can be opened in browser in near future, where all the build results can be seen)
- Cause of the build

Changes
 - Adds a new stage `Notify user` build pipeline
 - Added new env vars in seed-job/template.yml
 - Added `ccp/notifications` module
 - Added library modules for
   - email utilities (any module can use it and send emails)
   - OpenShift utility to get Jenkins authorization token from jenkins secret using oc
   - OpenShift utility to get Jenkins build information
- Added unittests for notifications and library modules

bamachrn committed:
Fixed all the registry endpoints to get proper value in GUI

navidshaikh committed:
Refactors base class for scanners and adds utilities

  The new refactored base class now has organized utilities for scanning.

navidshaikh committed:
Refactors individual scan invokers and scan runner

  also removed output file names config altogether from settings.

navidshaikh committed:
Updates scanner functional tests to use individual invoker classes

   Now result file name is used from individual invoker classes.
   Enabled pipeline scanner result file test.
   Reduced amount of sleep time while expecting results to appear.

navidshaikh committed:
Fixes retrieving image identifier key name from job data

navidshaikh committed:
Unifies scanner variable names across invoker classes

navidshaikh committed:
Updates base and subsclass scan and run methods

  Also fixes scnanner_name --> scanner variable

navidshaikh committed:
Adds missing mount point to mount object for pipeline scanner

navidshaikh committed:
Fixes processing result on failure for pipeline scanner

navidshaikh committed:
Fixes processing output for misc-package-updates scanner

navidshaikh committed:
Corrects the installed name of rpm-verify scanner

navidshaikh committed:
Exports scanner execution status file with indent and sorted keys

navidshaikh committed:
Fixes processing output result format for pipeline scanner

navidshaikh committed:
Sets required env vars in base class instead of subclass

navidshaikh committed:
Improves default result processing of base class

navidshaikh committed:
Fixes imports in scanner functional tests

  Using the result file names of each scanner as fixtures and avoiding
  to import individual scanner invoker class in tests.

navidshaikh committed:
Show only image name in scanner output

  Earlier registry name along with random tag used scanning will
  show up in scanners output. This changeset adds a utility method
  to split the repository_name and only add image_name without tag
  in scanners output.

dharmit committed:
Fix settings for scanner worker to look for correct file

navidshaikh committed:
Merge pull request #570 from dharmit/fix-569

Fix settings for scanner worker to look for correct file

navidshaikh committed:
Uses devicemapper storage option for devcloud environments

  Fixes #567: We were using default storage driver overlay2 on devcloud
  environment, causing non writable mounts for pipeline-scanner.
  This changeset updates provisioning scripts to use devicemapper as
  storage driver when deployment environment is not in [production, pre-prod, staging].

bamachrn committed:
Merge pull request #372 from mohammedzee1000/11-23-2017_14-00-18_pipeline_data_rest_api

Adding api to the container pipeline service data.

mohammedzee1000 committed:
Fixing target file link addition in weekly scan.

Signed-off-by: Mohammed Zeeshan Ahmed <mohammed.zee1000@gmail.com>

navidshaikh committed:
Stops using separate project to check scanners results, use existing

  Now scanners tests will check for scanners results of projects
  built already via openshift tests.
  This simplifies the tests and overall running time.

mohammedzee1000 committed:
Removing unused import statement.

navidshaikh committed:
Merge pull request #573 from mohammedzee1000/04-16-2018_12-10-21_fix_weekly_scan_db_gc

Fixing target file link addition in weekly scan.

dharmit committed:
Merge pull request #571 from navidshaikh/fix-devicemapper

Uses devicemapper storage driver for devcloud environments

bamachrn committed:
Update openshift version to fix oc retrive logs in workers

dharmit committed:
Merge pull request #558 from navidshaikh/fix-550

Improves scanning module, better error handling and cleanup, DRY scan invokers

bamachrn committed:
Added port to public registry in ci

bamachrn committed:
Added sample for public registry in inventory sample

bamachrn committed:
Added intranet registry as preferance to pull images

bamachrn committed:
resotred hosts.sampl to make registry ui work

bamachrn committed:
Removing unnecessarry public registry

dharmit committed:
Merge pull request #520 from bamachrn/fix-registry-url

Fixed all the registry endpoints to get proper value in GUI

mohammedzee1000 committed:
Refactoring index CI to a more modular approach.

Index CI is being re-written to allow modularity
and an easier way to use the existing validation.

mohammedzee1000 committed:
Fixing pep8 issues in code.

mohammedzee1000 committed:
Fixed bug with overall success and removed DependsOnValidation.

Depends on validation moved out of this patch.

bamachrn committed:
Removed gitignores to get all the files in repo

bamachrn committed:
Removed gitignores to get all the files in repo

dharmit committed:
Merge pull request #584 from bamachrn/pre-build-git-push

Removed gitignores to get all the files in repo

mohammedzee1000 committed:
Removing state related stuff from PR, and it will be part of pr 2.

mohammedzee1000 committed:
Changing clone function to directly clone to destination.

mohammedzee1000 committed:
Added code to enable storage of state accross run.

mohammedzee1000 committed:
Adding constants for state data.

mohammedzee1000 committed:
Adding state initialization and cleanup to the engine.

mohammedzee1000 committed:
Updating state logic for storage of state and keys.

mohammedzee1000 committed:
Correcting logic for yaml dump and adding 2 more functions.

mohammedzee1000 committed:
Adding base classes for stateful validations.

mohammedzee1000 committed:
Adding UniqueEntryValidator as first stateful validatior.

mohammedzee1000 committed:
Enabling UniqueEntryValidator by default.

mohammedzee1000 committed:
Removed uuid as its no longer needed here and fixed import.

mohammedzee1000 committed:
Adding constants to be used by validators.

mohammedzee1000 committed:
Updating clone function to strip leading //

mohammedzee1000 committed:
Removing unnessasary ".

mohammedzee1000 committed:
Updating and renaming cloned validator to allow all post clone validations.

mohammedzee1000 committed:
Wiring the value validators into the engine.

dharmit committed:
Parameterize container index repo and branch

Filters configuration files during execution of rpm verify scan

Invokes python script from slave container for scan

  A new image is also referenced which contains the mentioned
  python script for yum updates scan.

  In scan stage, the scan container is created with volume share
  of /opt/scanning dir. This dir is present in the slave container.
  This dir contains the yum updates scan script.
  The volume is made available in scan container, and the respective
  script is invoked using docker exec in matching scan stage.

[tmp] Serves image from temp registry until image appears on r.c.o

Removes extra whitespaces while redirecting stderr to stdout

DEBUG: Check the contents of /opt/ dir in slave container

Uses only docker run approach for scan

Overrides the work dir while creating the scan container

DEBUG: Check the docker run with entrypoint approach

mohammedzee1000 committed:
Adding value_validation.

mohammedzee1000 committed:
Adding a print function to utils.

mohammedzee1000 committed:
Fixing engine.py based on review.

- Adding error message and flow in case we failed to fetch data from index file.
- Adding docstrings.
- Fixing imports

mohammedzee1000 committed:
Updating utils as per review by navidshaikh

- Fixing imports
- Moved return True out of exception
- load and dump yaml now returns error message that can be used by callee.

mohammedzee1000 committed:
Fixing load_yaml call to handle err and display it to user.

mohammedzee1000 committed:
Fixing basevalidation.py as per comments by navidshaikh

- Adding missing docstrings.
- Adding default value UNKNOWN to field_name.
- Fixing typo.

mohammedzee1000 committed:
Fixing schema_validation as per comments by navidshaikh.

- Adding docstrings
- Changing dicts to dictionaries
- Fixing consistency of Id im comments

mohammedzee1000 committed:
Fixing pep8 issues.

navidshaikh committed:
Merge pull request #582 from mohammedzee1000/index_ci_refac

Refactoring index CI to a more modular approach.

mohammedzee1000 committed:
Fixing pep8 issues.

mohammedzee1000 committed:
Fixing issues based on comments by @navidshaikh

- Fixing all imports
- Adding missing docstrings
- Fixed formatting issues with code.

mohammedzee1000 committed:
Adding Pre-build schema validation.

mohammedzee1000 committed:
Adding validation for target-file.

mohammedzee1000 committed:
Adding CCCP Yaml Validator.

mohammedzee1000 committed:
Adding new validations to the index ci.

- Updating CCCP Exists Validator to use new CCCPValidator base.
- Adding prebuild exists and Job ID matching validation.
- Target file validation now skips if prebuild is present.

mohammedzee1000 committed:
Adding newly added validations to validation list.

mohammedzee1000 committed:
Fixing pep8 issues.

mohammedzee1000 committed:
Adding missing docstrings and comments.

mohammedzee1000 committed:
Fixing space in comment.

mohammedzee1000 committed:
Adding tests for new index CI.

- Adding basic framework.
- Adding tests for projects validation and id validation.

navidshaikh committed:
Merge pull request #585 from mohammedzee1000/index_ci_refac_2

This is the 2nd patch of the index ci refactor series.

mohammedzee1000 committed:
Fixing test order and adding more tests.

mohammedzee1000 committed:
Adding missing docstrings.

mohammedzee1000 committed:
Adding few more schema validation tests and fixing some exisiting ones.

mohammedzee1000 committed:
Adding build context schema validatation.

mohammedzee1000 committed:
Adding pre-build validaion tests and fixing issues found in validator.

mohammedzee1000 committed:
Altered the way state is implemented.

State can now be maintained by each callee
individually and people can juggle around
state objects.

mohammedzee1000 committed:
Adding variable to allow state to be sent as validation data.

mohammedzee1000 committed:
Altered validators and engine to use new state format.

Stateful validator updated to read state from validation data
Engine initialized state object to share with the validators.

navidshaikh committed:
Merge pull request #586 from mohammedzee1000/index_ci_refac_3

Adding remaining baseline validations to index ci

navidshaikh committed:
Adds documentation about configuring DaemonSet

mohammedzee1000 committed:
Removing unnessasary functions.

mohammedzee1000 committed:
Adding UniqueEntryValidation tests.

mohammedzee1000 committed:
Adding tests for git clone validation.

mohammedzee1000 committed:
Fixing pep8 issues.

navidshaikh committed:
Fixes validating projects against db from index in weekly scan

  We must validate project entry db against entry in index before
  putting the weekly scan job on tube.
  Handles the DoesNotExist exception raised by django, if entry in index
  does not match in projects table of db.

navidshaikh committed:
Merge pull request #588 from mohammedzee1000/index_ci_state_update

Provisions having separate instance of state object

mohammedzee1000 committed:
Updating state to be able to handle environment for git.

mohammedzee1000 committed:
Adding environment setup to OptionalCloneValidator as it does git clone.

mohammedzee1000 committed:
Adding remaining validations for git clone.

mohammedzee1000 committed:
Altering how CI index CI works with paths containing gitpath.

Git path can start with "/" which is problem for os.path.join

mohammedzee1000 committed:
Adding a mock directory to state for tests and any other rough items.

mohammedzee1000 committed:
Adding base class with common function for file exists validations.

mohammedzee1000 committed:
Adding tests for cccp yaml exists validations.

mohammedzee1000 committed:
Adding tests for validating TargetFile validation.

mohammedzee1000 committed:
Renaming FileBaseTest, adding PreBuildValidation tests, and fixing.

mohammedzee1000 committed:
Adding tests for job id in cccp yaml validation and fixing issues.

mohammedzee1000 committed:
Removing unnessasary function.

mohammedzee1000 committed:
Fixing <= to ==.

mohammedzee1000 committed:
Renamed functions that had missed rename in git url schema tests.

mohammedzee1000 committed:
Fixing names in git path validation.

navidshaikh committed:
Sends email for weekly scan only if scanners execution status=true

  In case of scanners failed to execute, prevent empty emails to be sent.
  Also corrects the build status to `failed` if scanners failed to execute.

navidshaikh committed:
Merge pull request #590 from CentOS/fix-589

Fixes validating projects against db from index in weekly scan

mohammedzee1000 committed:
Updating container index ci engine to allow state to be passed.

The engine can now be given a state object to use. This enables
user to create the state object outside, initialize extra stuff,
and then pass it to the engine to use.

Also adding functional tests.

navidshaikh committed:
Merge pull request #587 from mohammedzee1000/index_ci_refac_4

Adding tests for container index CI

mohammedzee1000 committed:
Engine can now skip validations, by specifying None.

Signed-off-by: Mohammed Zeeshan Ahmed <mohammed.zee1000@gmail.com>

mohammedzee1000 committed:
Removed unused imports.

mohammedzee1000 committed:
Adding functional tests for index ci and removing debug statements.

mohammedzee1000 committed:
Changed the way to to specify no validations.

mohammedzee1000 committed:
Adding return if cccp yaml data failed to load.

mohammedzee1000 committed:
Updating tests to use the new way of no validators.

mohammedzee1000 committed:
Adding ability to manipulate parsers to init.

mohammedzee1000 committed:
Adding index ci functional tests for command line invokation.

The tests will now need to be invoked with sudo privilages.

mohammedzee1000 committed:
Changing toe check_call.

mohammedzee1000 committed:
Rewiring index ci to run the new validations.

bamachrn committed:
Added jenkins file for ci-functional test

bamachrn committed:
Added job template for CI

bamachrn committed:
Added ci trigger point

mohammedzee1000 committed:
Removing references that pull different branch.

mohammedzee1000 committed:
Removing pythonpath as it can be added while calling the code.

mohammedzee1000 committed:
Altering script to use python path.

navidshaikh committed:
Merge pull request #591 from mohammedzee1000/index_ci_refac_5

Adding more tests for the index CI

mohammedzee1000 committed:
Fixing the way you can skip entire set of tests.

mohammedzee1000 committed:
Fixing the parameters according to changes in above commit.

mohammedzee1000 committed:
Fixing the tests to work with new way of skipping validations.

mohammedzee1000 committed:
Fixed based om comments by @navidshaikh

bamachrn committed:
Added approvals for running the JSON checks

bamachrn committed:
Added function to remove repeated functionalities

mohammedzee1000 committed:
Setting pythonpath before running the CI.

The pythonpath needs to be set to root of service code
or imports will fail.

bamachrn committed:
Added parallel check for both success and failure

bamachrn committed:
Added assertions for success and fail cases

mohammedzee1000 committed:
Updating yaml loader to use base loader.

navidshaikh committed:
Merge pull request #595 from mohammedzee1000/07-17-2018_14-42-38_pipeline_ci_python_path

Setting pythonpath before running the CI.

mohammedzee1000 committed:
Removing yaml loader, setting it to default.

mohammedzee1000 committed:
Fixing the help test for parameter.

mohammedzee1000 committed:
Removing unneeded commas in test data.

bamachrn committed:
Added nfs setup provisions

mohammedzee1000 committed:
Set python path for ci run.

mohammedzee1000 committed:
Adding debug messages.

bamachrn committed:
Merge pull request #593 from mohammedzee1000/index_ci_refac_6

Adding init for new index ci and adding tests for the same

bamachrn committed:
Merge pull request #594 from mohammedzee1000/index_ci_refac_rewire_index_ci

Rewiring index ci to run the new validations.

mohammedzee1000 committed:
Adding more explicit setting of pythonpath for index ci.

navidshaikh committed:
Merge pull request #596 from mohammedzee1000/index_ci_pythonpath

Adding more explicit setting of pythonpath for index ci.

mohammedzee1000 committed:
Fixing validator load conditions.

Still need to figure out how this slipped pass.

mohammedzee1000 committed:
Fixing index ci to skip cccp yaml validation if prebuild is specified.

mohammedzee1000 committed:
Fixing the print function to not print state.

bamachrn committed:
Added ansible-controller

bamachrn committed:
Added common for all

bamachrn committed:
Allow all requests from the cluster nodes

mohammedzee1000 committed:
Adding logic to skip template file.

mohammedzee1000 committed:
Fixing prebuild validation to check prebuild script relative to context.

mohammedzee1000 committed:
Fixing index ci tests for pre-build validation.

bamachrn committed:
Fixed nfs export for PV

bamachrn committed:
Added Docker distribution to NFS node

bamachrn committed:
Deploy Openshift cluster

mohammedzee1000 committed:
Temporarily adding skip for centos base image.

bamachrn committed:
Deploy jenkins with persistent volume

bamachrn committed:
Added script approval for running json checks

mohammedzee1000 committed:
Disabling stringfield validation on desired tag.

mohammedzee1000 committed:
Adding library images to list of those where job id match is skipped.

mohammedzee1000 committed:
Adding few exceptions to index ci until they can be sorted.

bamachrn committed:
Merge pull request #601 from mohammedzee1000/index_ci_fixes

Adding few exceptions to index ci until they can be sorted.

mohammedzee1000 committed:
Making the skip of fabric8-ui images more loose.

bamachrn committed:
Merge pull request #602 from mohammedzee1000/index_ci_fixes

Making the skip of fabric8-ui images more loose.

mohammedzee1000 committed:
Adding fabric8 namespace as well.

navidshaikh committed:
Adds notification module for sending emails to user

  - Added a notification module
  - Added a notify user stage
  - Updated seed-job/template.yaml to clone the pipeline_repo
  - Added respective parameters and env vars in seed-job/template.yaml

bamachrn committed:
Added hosts file for OpenShift 3.9

bamachrn committed:
Added ci trigger on CICO

bamachrn committed:
Added trigger script for CI

bamachrn committed:
Set log level to error

mohammedzee1000 committed:
Removing old index ci in favour of new one.

This PR removes the old index ci along with tests associated
with the old index ci.

bamachrn committed:
Changed ci index repo to upstream

bamachrn committed:
Run CI pipeline and release nodes after done

bamachrn committed:
Move build config to use scm plugin instead of git

bamachrn committed:
Added openshift authentication

navidshaikh committed:
Merge pull request #604 from mohammedzee1000/delete_old_index_ci

Removing old index ci in favour of new one.

bamachrn committed:
Fetch CI logs

navidshaikh committed:
Get the build number of currently running job using env var

navidshaikh committed:
Adds accessing cause of build details for update in source git repo case

  Accessing the details for cause of build build in case of
  update to source git repo, by checking the _class names instead
  of harcoding the indexes os response.

  Also added git url and git branch along with git sha, submitted to
  referenced source git repo.

navidshaikh committed:
Uses secret token from sa/jenkins

navidshaikh committed:
Restructures notify module to make it object oriented

  Addresses success case of build and groups all related utilities
  into respective classes.

navidshaikh committed:
Parameterizes the notification module invocation

  Adds the case for build failure
  Generates the jenkins pipeline job name as namespace-pipeline_name

navidshaikh committed:
References the pipeline-images/ccp-jenkins-slave container in template

  Reverts back the custom jenkins slave image reference.

navidshaikh committed:
Separates method to parse the API JSON response for jenkins job

  Adds a method to parse the REST API JSON response containing
  Jenkins job details.
  This is to unit test the JSON parsing.

navidshaikh committed:
Uses the ccp-openshift-slave image from pipeline-images namespace

 reverts back the personal namespaced image reference.

navidshaikh committed:
Adds email module  in ccp library utilities

  Module has Email class with email utilities.

navidshaikh committed:
Adds openshift module in ccp library utilities

 Module has openshift operations related utilities.

 Added get_token and BuildInfo class.

navidshaikh committed:
References email and buildinfo utilities modules from ccp.lib

  Uses buildinfo module to get the needed details for notifications.
  Uses email module to send emails.

navidshaikh committed:
Adds SMTP_SERVER and FROM_ADDRESS to inline Jenkinsfile env vars

  This is added so that, Jenkins job build details has mentioned
  variables. Later on these variables, are retrieved from Jenkins job
  REST APIs and used in notificaiton module.

  Invokes notification module in notify stage with only required
  parameters to call REST APIs.

navidshaikh committed:
Uses list comprehension to make cc options string for mail command

navidshaikh committed:
Raises exception upon failure to open Jenkins job build info URL

  Also fixes typo in command for getting the secrets command.
  Uses %s string formatting for oc command having double curly
  brackets as using formatting will convert double curly brackets
  into single.

navidshaikh committed:
Adds unittests for notifications.notify module

navidshaikh committed:
Adds unittests for ccp.lib.openshift module

navidshaikh committed:
Adds build failure email body template

  Also fixes the cause of build line formatting and other typos.

navidshaikh committed:
Adds fetching cause of build for first build of the project

  The first build of the project starts by default, added
  provision for same.

navidshaikh committed:
Adds https:// before repository URL

navidshaikh committed:
Adds container-index link in the footer of email

navidshaikh committed:
Fixes processing status of the build

 - removes mistakenly added python scripts

navidshaikh committed:
Parses upstream/parent project rebuilt and manual trigger cause of build

  Adds ability to identify cause of build for cases
   - when upstream/parent project is rebuilt
   - when the build has started by admin manually

navidshaikh committed:
Fixes a typo and improves formatting for email body text

 also updates respective unittests

navidshaikh committed:
Adds script to run unittests in Centos CI infrastructure

 This patch adds script which invokes CI job at cico for
 running unittests.

navidshaikh committed:
Updates the cause of build message for build configuration update

The possible causes of builds (as per current implementation of OpenShift based service) are:

1 -  First build of container (after indexing)
2 -  Git commit $SHA [..]
3 -  Upstream/parent project is rebuilt
4 -  Project configs (like desired-tag, emails) have updated
5 -  Manually triggered by admin

We have implemented 1, 2, 3 and 5.

For 4:
OpenShift / oc can identify only the new buildConfigs created -- meaning --> if buildConfig is updated for an existing project, OpenShift does not trigger the build by itself.

How we have handled this for builds?
A - We have implemented logic in our index reader. After reading index we identify, if a project configs are updated, we trigger the build using `oc start-build` command explicitly.

On the receiving end (i.e. notification module), when checks Jenkins job information, it sees, the build (whose configs are updated) is manually triggerred (since we ran oc `start-build` refer A)

Thus - programmaticaly, we have similar means for cases 4 and 5, i.e. When project configs are updated, the jenkins identifies it as manual trigger,
and there is no identifier for same in its current form. (given we are parsing Jenkins APIs to find cause of build)

I have kept the string "Manually triggerred by build." for case 5 (and implicitly 4).

For production scenario, I am sure, we are going to get case 4 very often as compare to case 5 (one of our team member manually starting the build).
Given this, do we want to change that message?

Keeping "Update to build configurations of container"

navidshaikh committed:
Adds unittests for 5 cases of cause of builds

  - Adds unittests for parsing different Jenkins build job API response
  - Tests every implemented cause of build case
  - Asserts exception is raised when invalid response is given

  Also updates other unit tests docstrings to represent
  the module name, as the docstring is shown in
  the nosetests output. eg: nosetests -w . -vv tests/

navidshaikh commented:
ping @bamachrn @dharmit @mohammedzee1000 : this is ready for review.

navidshaikh commented:
Unit tests added in this PR: 

```
➜ nosetests -w . -v  tests/test_00_unit/test_02_ccplib/test_01_openshift.py 
ccp.lib.openshift: Test processing Jenkins pipeline job name ... ok
ccp.lib.openshift: Test parsing cause of build case-1 (Git commit) ... ok
cccp.lib.openshift: Test parsing cause of build case-2 (parent proj) ... ok
cccp.lib.openshift: Test parsing cause of build case-3 (first build) ... ok
cccp.lib.openshift: Test parsing cause of build case-4 (config update) ... ok
cccp.lib.openshift: Test parsing cause of build case-5 (failure) ... ok
ccp.lib.openshift: Test parsing build info from ... ok

----------------------------------------------------------------------
Ran 7 tests in 0.001s

OK
➜  nosetests -w . -v  tests/test_00_unit/test_01_notifications               
Notifications: Tests processing body of email ... ok
Notitications: Tests processing subject of email ... ok

----------------------------------------------------------------------
Ran 2 tests in 0.000s

OK
```

bamachrn committed:
Brought jenkins file to inline

bamachrn committed:
Added pre-requisite for deployment

bamachrn committed:
Wait for seed job to complete before ci starts

bamachrn committed:
Merge branch 'openshift' of github.com:bamachrn/ccp-openshift into openshift

navidshaikh committed:
Updates default values of git branch, pipeline repo and index

  to current remote of service code base

bamachrn committed:
Removed old code base in faovour of new one

bamachrn committed:
Separate success job

bamachrn committed:
Added failure check pipeline

navidshaikh committed:
Adds BaseNotify class and inherits it in BuildNotify class

 - Adds a base.py class having BaseNotify base class for notification
   module

 - Updated the name of Notify class to BuildNotify class

 - Defines all the constants strings and templates in base class

navidshaikh committed:
Adds sphinx style docstring to ccp/notifications/notify.py file

 Also converts build_number to str before using it in URL.

navidshaikh committed:
Relays PIPELINE_REPO and PIPELINE_BRANCH param processing in index reader

  Added two new parameters in buildtemplate.yml, it needs to be also
  passed to template.yml, because the service code is referenced in template.yml
  and the right branch and repo needs to be checked out before referencing the code.

navidshaikh committed:
Adds [registry.centos.org] in email subject at the start

navidshaikh committed:
Fixes typos in email body

navidshaikh commented:
@mohammedzee1000 @bamachrn @dharmit : This is ready to review.

navidshaikh committed:
Updates the unit tests with recent changes in text processing

 Since we updated some email subject and body text

navidshaikh commented:
@mohammedzee1000 @bamachrn @dharmit : This is ready to review.

bamachrn committed:
Functional CI work flow diagram

This diagram explains the flow of functional CI

bamachrn committed:
Added documentation for functional tests

navidshaikh committed:
Updates the email subject and removes redundant words

 Also updates respective unit tests

navidshaikh committed:
Converts pipeline name to lower case

 Fix OSIO 1522

 Since OpenShift doesn't allow the value of `metadata.name` to have uppercase chars.

 This commit also adds a unit test case.

navidshaikh committed:
Ensures the pipeline name confronts to allowed pattern

 Added a regular expression to validate the pipeline name generated
 for using in template's metadata.name field. oc imposes certain guidelines
 for same and provides the regular expression in error/help message to be
 used for validation in case it found some invalid string.

 We are using same regular expression as provided by oc to validate
 the pipeline name while processing index and generating build configs
 from it.

 Added a custom exception, which will be raised if the populated string
 is invalid.

 Added a unit test as well to test the functionality.

navidshaikh committed:
Sends PIPELINE_NAME from template to notification module

 Since PIPELINE_NAME parameter's value processing is updated and
 the same value should be used in notifications module to formulate
 the URL to get the Jenkins build info.

bamachrn commented on ccp/index_reader.py:
```diff
@@ -126,9 +136,27 @@ def load_project_entry(self, entry):
+        pipeline_name_regex = ("^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]"
+                               "([-a-z0-9]*[a-z0-9])?)*$")
+        match = re.match(pipeline_name_regex, pipeline_name)
+
```
we need to add check to `container-index` as well for this

bamachrn commented on ccp/index_reader.py:
```diff
@@ -219,7 +252,9 @@ def __init__(self, registry_url, namespace, from_address, smtp_server):
+-p SMTP_SERVER={smtp_server} \
+-p PIPELINE_REPO={pipeline_repo} \
+-p PIPELINE_BRANCH={pipeline_branch}"""
 
```
can't we fix the indentation here? it looks confusing while going through parametes

bamachrn commented on ccp/lib/command.py:
```diff
@@ -0,0 +1,21 @@
+    :return: Command output
+    :rtype: str
+    :raises: subprocess.CalledProcessError
+    """
```
:+1:  nice documentation

navidshaikh committed:
Parameterizes ccp-openshift-slave image and references modules from image

  Added a parameter to specify the ccp-openshift-slave container image
  to be used.

  The ccp-openshift-slave container image will now contain the service
  source code.

  Added another parameter $PIPELINE_REPO_DIR, which would specify
  the source code path inside container, and the same will be referenced
  to invoke the needed modules from template.

navidshaikh committed:
Adds the pipeline service code in ccp-openshift-slave container

  The build context referenced in Dockerfile is root dir of
  service source code.

  The command to build the docker image is
  # root dir of the service source repo
  $ docker build -t $REGISTRY/pipeline-images/ccp-openshift-slave -f Dockerfiles/ccp-openshift-slave/Dockerfile .

bamachrn commented on ccp/lib/openshift.py:
```diff
@@ -0,0 +1,209 @@
+{{ .name }} {{end}}' | xargs -n 1 oc get secret --template=\
+'{{ if .data.token }}{{ .data.token }}{{end}}' | head -n 1 | \
+base64 -d -""" % service_account
+    # not using format in above command, as it will convert
```
may be we can use the same library for all the cases where we need to access jenkins

bamachrn commented on ccp/notifications/base.py:
```diff
@@ -0,0 +1,38 @@
+        self.build_failure_body = """\
+{0:<30}{1}
+{2:<30}{3}"""
+
```
can we add a bit of documentation, what we are doing here?

bamachrn commented on ccp/notifications/notify.py:
```diff
@@ -0,0 +1,137 @@
+            body = self.build_failure_body.format(
+                "Build Status:", "Failure",
+                "Cause of build:", cause)
+
```
:+1:

bamachrn commented on seed-job/template.yaml:
```diff
@@ -33,8 +33,8 @@ objects:
-                    name: 'jnlp',
-                    image: 'registry.centos.org/pipeline-images/ccp-openshift-slave',
+                    name: 'ccp-openshift-slave',
+                    image: ${CCP_OPENSHIFT_SLAVE_IMAGE},
```
build template has the image fixed to `r.c.o/pipeline_images/ccp-openshift-slave ` I beleive we need to update that as well.

bamachrn commented on seed-job/template.yaml:
```diff
@@ -169,6 +170,12 @@ objects:
+          - name: SMTP_SERVER
+            value: ${SMTP_SERVER}
+          - name: PIPELINE_REPO_DIR
+            value: ${PIPELINE_REPO_DIR}
```
I did not get where this paramaters is getting filled in

navidshaikh committed:
Parameterizes CCP_OPENSHIFT_SLAVE for seed-job/build_template.yaml

 Adds CCP_OPENSHIFT_SLAVE parameter which would specify the slave container
 image to be used.

 Updates the default value of PIPELINE_REPO_DIR variable to
 /opt/ccp-openshift , this is directory where service source code
 is added in slave container.

 Invokes the python modules from PIPELINE_REPO_DIR.

navidshaikh committed:
Relays CCP_OPENSHIFT_SLAVE_IMAGE parameter from buildtemplate to template

  CCP_OPENSHIFT_SLAVE_IMAGE parameter is relayed via index_reader to
  template.yml from buildtemplate.yml

  Since index_reader parameterizes the template.yml, it needs the value
  for same.

navidshaikh committed:
Changes dir to $PIPELINE_REPO_DIR before invoking service module

 Follows the convention from buildtemplate.yml when it invokes the
 index reader

navidshaikh committed:
Adds quotes around parameter referenced for image field's value

  Since slave image's value is referenced via variable,
  adds quotes around it.

navidshaikh committed:
Parameterizes CCP_OPENSHIFT_SLAVE for weekly-scan/template.yaml

 As done in other templates

bamachrn commented on ccp/lib/openshift.py:
```diff
@@ -0,0 +1,209 @@
+        self.required_fields = required_fields
+
+    def get_url(self, url, token, context=None):
+        """
```
Here we are actually fetching the job details, may be we can rename it to something else?

bamachrn commented on ccp/lib/openshift.py:
```diff
@@ -0,0 +1,209 @@
+                        build_info[param["name"]] = param["value"]
+        except KeyError as e:
+            print ("Error parsing build details.")
+            raise(e)
```
Here we are raising the error this means it will fail the pipeline right?

navidshaikh replied:
Yes - There is no change in the behavior. The seed job was failing if it failed to parse even a single project and the same is happening here with a raise.

I remember our discussion to change the behavior of seed-job to move on and create other jobs upon failing to process a job in sequence, however, that behavior change should be scoped in another PR.

navidshaikh committed:
Adds quotes around parameter referenced for image field's value

 Since slave image's value is referenced via variable,
 adds quotes around it.

navidshaikh committed:
Sets PYTHONPATH before invoking the notification module

 in Notify user stage

bamachrn committed:
Getting CI logs for both success and failure

bamachrn committed:
Updated README with both success and failure logs

navidshaikh committed:
Updates the name fields value to jnlp for ContainerTemplate

  for all the templates

navidshaikh committed:
Uses absolute path of /bin/email binary

 Since we are checking out the dir to PIPELINE_REPO_DIR dir in
 templates.

navidshaikh committed:
Removes changing dir to PIPELINE_REPO_DIR in Notify user stage

 Since it changes the dir and then mailx command path is used
 as relative and subprocess can't find the mailx command.

navidshaikh committed:
Installs postfix package as emailing complains missing sendmail binary

navidshaikh committed:
Adds another run_cmd2 in lib

navidshaikh committed:
Starts postfix

bamachrn committed:
Added comments to both pipelines

dharmit committed:
Merge pull request #608 from bamachrn/openshift

OpenShift based new architecture and its functional CI

bamachrn committed:
Merge pull request #613 from navidshaikh/os-unittests

Adds script to run unit tests in Centos CI infrastructure
